### ISSUE TITLE
Race condition prevents Node.js process from exiting

### ISSUE THREAD

**Issue by jamesshore** [socketio/socket.io-client#935]
When using socket.io-client in Node, if the server is closed too soon after the client connects, disconnect() doesn't work properly and the process never exits.

To reproduce, use this code with Node v5.4.0 and socket.io-client@1.4.3:

```javascript
(function() {
  "use strict";

  var clientIo = require("socket.io-client");
  var http = require("http");
  var io = require('socket.io');

  var httpServer;
  var server;

  var PORT = 5020;

  startServer(function() {
    console.log("SERVER STARTED");

    var client = clientIo("http://localhost:" + PORT);

    client.on('connect', function() {
      client.disconnect();
      stopServer(function() {
        console.log("COMPLETE! NODE SHOULD EXIT NOW.");
      });
    });
  });

  function startServer(callback) {
    httpServer = http.createServer();
    server = io(httpServer);
    httpServer.listen(PORT, callback);
  };

  function stopServer(callback) {
    httpServer.on("close", function() {
      console.log("SERVER CLOSED");
      callback();
    });

    server.close();
  };

}());
```

Output:

```sh
$ node spikes/socket.io/test.js 
SERVER STARTED
SERVER CLOSED
COMPLETE! NODE SHOULD EXIT NOW.
^C
```

If you introduce a delay in the stopServer() function, everything works fine:

```javascript
  function stopServer(callback) {
    httpServer.on("close", function() {
      console.log("SERVER CLOSED");
      callback();
    });

    setTimeout(function() {
      server.close();
    }, 500);
  };
```

Output:

```sh
$ node spikes/socket.io/test.js 
SERVER STARTED
SERVER CLOSED
COMPLETE! NODE SHOULD EXIT NOW.
$
```

**Comment by mspoulsen** [comment]
Hi James, 

I added two lines and now it should exit. I have no clue why though :)

```javascript

(function() {
    "use strict";

    var clientIo = require("socket.io-client");
    var http = require("http");
    var io = require('socket.io');

    var httpServer;
    var server;

    var PORT = 5020;

    startServer(function() {
        console.log("SERVER STARTED");

        // added options object
        var client = clientIo("http://localhost:" + PORT, { transports: ['websocket'] });

        client.on('connect', function() {
            client.disconnect();

            stopServer(function() {
                console.log("COMPLETE! NODE SHOULD EXIT NOW.");
            });
        });
    });

    function startServer(callback) {
        httpServer = http.createServer();
        server = io(httpServer);
        httpServer.listen(PORT, callback);
    }

    function stopServer(callback) {
        // added server close
        server.close();         
        httpServer.on("close", function() {
            console.log("SERVER CLOSED");
            callback();
        });

    }

}());

```

**Comment by nkzawa** [comment]
Should be fixed on v1.4.5. Related https://github.com/socketio/socket.io/issues/2069
